test/service/test_response.py

```diff
@@ -2035,24 +2042,63 @@ def test_summary_filter_none(self):
 
 class TestUnpopulatedIndexResponse(WebServiceTestCase):
 
+    @classmethod
+    def bundles(cls) -> List[BundleFQID]:
+        return []
+
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
-        cls.index_service.create_indices(cls.catalog)
+        cls._setup_indices()
+        cls.maxDiff = None
 
     @classmethod
     def tearDownClass(cls):
-        cls.index_service.delete_indices(cls.catalog)
+        cls._teardown_indices()
         super().tearDownClass()
 
+    def facets(self) -> Sequence[str]:
+        return self.app_module.app.facets
+
+    def entity_types(self) -> List[str]:
+        return [
+            entity_type for entity_type in self.index_service.entity_types(self.catalog)
+            if entity_type != 'cell_suspensions'
+        ]
+
     def test_empty_response(self):
-        url = self.base_url + "/index/projects"
-        response = requests.get(url)
-        response.raise_for_status()
-        response = response.json()
-        self.assertEqual([], response['hits'])
-        self.assertEqual({None}, set(response['pagination'].values()))
-        self.assertEqual({}, response['termFacets'])
+        empty_term_facets = {
+            facet: {'terms': [], 'total': 0, 'type': 'terms'}
+            for facet in self.facets()
+        }
+        for entity_type in self.entity_types():
+            with self.subTest(entity_type=entity_type):
+                response = requests.get(url=furl(url=self.base_url,
+                                                 path=('index', entity_type)).url)
+                response.raise_for_status()
+                response = response.json()
+                self.assertEqual([], response['hits'])
+                self.assertEqual(0, response['pagination']['count'])
+                self.assertEqual(0, response['pagination']['total'])
+                self.assertTrue(response['termFacets'])
+                for facet, body in response['termFacets'].items():
+                    self.assertEqual(empty_term_facets[facet], body)
+
+    def test_sorted_responses(self):
+
+        # Due to es_type not being defined for nested facets we remove them from
+        # the tests as they are known to product errors.
+        # see https://github.com/DataBiosphere/azul/issues/2621
+        sortable_facets = self.facets()
```

Comment:
Depending how `self.app_module.app.facets` is implemented, this may have the unintended side-effect of modifying it. Copy or use guarded set comprehension. In fact, Python gives you big fat warning:

![image](https://user-images.githubusercontent.com/5143256/102390837-c8fc9d00-3f89-11eb-9985-8efaae9b6ea2.png)

Does your PyCharm not show these?

Reply:
It did, i overlooked it, switched to guarded comprehension.